libslide/src/grammar.rs

```diff
@@ -230,14 +271,28 @@ impl fmt::Display for UnaryOperator {
     }
 }
 
-#[derive(PartialEq, Clone, Hash)]
+#[derive(PartialEq, Clone, Hash, Debug)]
 pub struct UnaryExpr<E: Expression> {
     pub op: UnaryOperator,
     pub rhs: Rc<E>,
 }
 
-impl<E: Expression> fmt::Display for UnaryExpr<E> {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{}{}", self.op.to_string(), self.rhs.to_string(),)
-    }
+macro_rules! display_unary_expr {
+    (<$expr:ident>) => {
+        impl fmt::Display for UnaryExpr<$expr> {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                let mut result = String::with_capacity(128);
+                use $expr::*;
+                let format_arg = |arg: &Rc<$expr>| match arg.as_ref() {
+                    BinaryExpr(l) => format!("({})", l),
+                    expr => expr.to_string(),
+                };
+                result.push_str(&format!("{}{}", self.op, format_arg(&self.rhs)));
+                f.write_str(&result)
+            }
+        }
+    };
 }
```

Comment:
I don;t understand why we are matching with a BinaryExpr in a unary Expr. Can you explain?

Reply:
A unary expression of the form `(- (+ 1 2))` must be written as `-(1 + 2)`. We match against a bin expr for the above reason. If the argument for the unary expression is a binary expression we have to wrap the binary expression in parentheses, otherwise it's incorrect (would be `-1 + 2`)

Reply:
gotcha